[Draft] Add docs on working with Pandera for data validation (#5142)
kedro-org/kedro: docs/integrations-and-plugins/pandera.md

Adding Pandera to a Kedro project enables you to catch data quality issues early in your pipeline. For example, you can validate that CSV columns have the correct types and enforce business rules (like "price must be positive"). When collaborating with others on a Kedro project, Pandera schemas serve as documentation of your data contracts.

## Prerequisites

You will need the following:

- A working Kedro project. The examples in this document assume the `spaceflights-pandas` starter. If you're unfamiliar with the Spaceflights project, check out [our tutorial](../tutorials/spaceflights_tutorial.md).
- Pandera installed into the project.

To set yourself up, create a new Kedro project:

```bash
kedro new --starter=spaceflights-pandas --name spaceflights-pandera
```

Navigate to your project directory and add Pandera with pandas support to your `requirements.txt`:

```bash
pandera[pandas]>=0.18.0
```

Install the project dependencies:

```bash
uv pip install -r requirements.txt
```

> ## Use cases

Reviewer: ```suggestion
## Integration Options
```

> # How to add Pandera data validation to your Kedro workflow

Reviewer 2: ```suggestion
# How to validate data using Pandera in your Kedro workflow
```
Maybe? I feel like it emphasizes validating data as the end goal more than adding a Pandera integration, but could go either way.

> [Pandera](https://pandera.readthedocs.io/) is an open-source data validation library for pandas DataFrames that enables schema validation with expressive and flexible checks. It helps ensure data quality by validating types, ranges, business rules, and statistical properties of your datasets.

Reviewer 2: Not just pandas.

> Adding Pandera to a Kedro project enables you to catch data quality issues early in your pipeline. For example, you can validate that CSV columns have the correct types, enforce business rules (like "price must be positive"), and automatically coerce string data to appropriate types. When collaborating with others on a Kedro project, Pandera schemas serve as documentation of your data contracts.

Reviewer 3: I would against promoting this pattern while this is possible. The I/O are usually define in DataCatalog, if the logic start leaking to hooks/data validation this makes it harder to debug. It also cause potentially problem, i.e. if data validation is not enable, the pipeline will be incorrect because of not casting the type.

Author: you mean the “coerce string data” part right?. That makes sense. I’ll also remove the example I share later on along with the notes. You’re right; pandera should focus on validation and raising errors, not casting or changing the data! thanks

Reviewer 3: exactly

## Prerequisites

You will need the following:

- A working Kedro project. The examples in this document assume the `spaceflights-pandas` starter. If you're unfamiliar with the Spaceflights project, check out [our tutorial](../tutorials/spaceflights_tutorial.md).
- Pandera installed into the project.

To set yourself up, create a new Kedro project:

```bash
kedro new --starter=spaceflights-pandas --name spaceflights-pandera
```

> Navigate to your project directory and add Pandera with pandas support to your requirements.txt:

Reviewer 2: ```suggestion
Navigate to your project directory and add Pandera with pandas support to your `requirements.txt`:
```

or

```suggestion
Navigate to your project directory and add Pandera with pandas support to your requirements file:
```

```bash
pandera[pandas]>=0.18.0
```

Install the project dependencies:

```bash
uv pip install -r requirements.txt
```

> 
> 
> !!! note
>     Although Pandera works best when working with tabular data pipelines, you can validate any pandas DataFrame in your Kedro workflow, not just ML/AI pipelines.

Reviewer 2: ```suggestion
```
I don't understand the purpose of this note? What are you trying to communicate?

## Simple use cases

> This section explains how you can use Pandera to validate your Kedro datasets in a straightforward way.

Reviewer 2: ```suggestion
This section explains how you can use Pandera to validate your Kedro datasets.
```
Too wordy?

    strict=False,
    coerce=True,
    name="reviews_raw",
)
```

!!! tip
    **Alternative schema style:**
    Instead of defining schemas as `DataFrameSchema` objects, Pandera also supports a Pydantic-like API with `DataFrameModel` classes and type annotations.
    This can give you IDE/type-checking support. See the [Typed Schemas guide](https://pandera.readthedocs.io/en/stable/schema_models.html).

!!! info
    Use `coerce=True` to automatically convert types during validation. This is essential for CSV data where columns are initially loaded as strings.

    For nullable integer columns, use `float` type instead of `int` because pandas represents nullable integers as float64 (since NaN is a float value).

Next, create a hook to validate datasets. Create `src/spaceflights_pandera/hooks/validation.py`:

```python
"""Kedro hooks for Pandera validation integration."""
import logging
from typing import Any, Dict

import pandas as pd
from kedro.framework.hooks import hook_impl
from kedro.io import DataCatalog
from pandera.errors import SchemaError, SchemaErrors

from ..schemas.raw import companies_schema, shuttles_schema, reviews_schema

logger = logging.getLogger(__name__)

> class PanderaValidationHooks:

Reviewer 2: I'm OK with it either way; I feel like use of "Hooks" as the class name suffix is pretty common.

    """Kedro hooks that integrate Pandera validation."""

    def __init__(self):
        # Map catalog dataset names to their validation schemas

>         self.raw_schemas = {

Reviewer 2: Feels very weird that you're defining your schemas in the hook; seems like it breaks the software engineering principles that Kedro advocates for. I know you mention somewhere below that you could store schemas in YAML, but maybe even for this example it makes sense to put a pointer to the schema in YAML or some other config?

Reviewer 4: I agree, we've struggled quite a bit with the validation logic. Either keep the validation logic close to the dataset, or keep it close to the node logic I'd say.

As part of our project, we went for the latter:

1. A small wrapper on top of Pandera to work both with Pandas and Spark
2. Decorators to validate input/output of functions

Some code:

Wrapper:

```python
"""
Temp. wrapper module to enable data quality checks defined independently of the underlying
library used to represent the data. Currently supports Spark and Pandas dataframes.

NOTE: Should be removed as soon as Pandera supports this.
"""

import inspect
import typing
from dataclasses import dataclass
from functools import wraps
from typing import Dict, List, Optional

import pandas as pd
import pandera.pandas as pa
import pandera.pyspark as pas
import pyspark.sql as ps
from pandera.decorators import _handle_schema_error

@dataclass
class Column:
    """Data class to represent a class agnostic Pandera Column."""

    type_: type
    checks: Optional[List] = None
    nullable: bool = True

@dataclass
class DataFrameSchema:
    """Data class to represent a class agnostic Pandera DataFrameSchema."""

    columns: Dict[str, Column]
    unique: Optional[List] = None
    strict: bool = False

    def build_for_type(self, type_) -> typing.Union[pas.DataFrameSchema, pa.DataFrameSchema]:
        # Build pandas version
        if type_ is pd.DataFrame:
            return pa.DataFrameSchema(
                columns={
                    name: pa.Column(col.type_, checks=col.checks, nullable=col.nullable)
                    for name, col in self.columns.items()
                },
                unique=self.unique,
                strict=self.strict,
            )

        # Build pyspark version
        if type_ is ps.DataFrame:
            return pas.DataFrameSchema(
                columns={
                    name: pas.Column(col.type_, checks=col.checks, nullable=col.nullable)
                    for name, col in self.columns.items()
                },
                unique=self.unique,
                strict=self.strict,
            )

        raise TypeError()

def check_input(schema: DataFrameSchema, df_name: str):
    """Decorator to validate output schema of decorated function.

    Args:
        schema: Schema to validate
        df_name (optional): name of output arg to validate
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            # Get type hints for the function
            type_hints = typing.get_type_hints(func)

            # Bind args and kwargs to parameter names
            sig = inspect.signature(func)
            bound = sig.bind(*args, **kwargs)
            bound.apply_defaults()

            value = bound.arguments.get(df_name)
            arg_type = type_hints.get(df_name)

            df_schema = schema.build_for_type(arg_type)

            if value is not None:
                try:
                    # breakpoint()
                    df_schema.validate(value, lazy=False)
                except pa.errors.SchemaError as e:
                    _handle_schema_error("check_input", func, df_schema, value, e)

            output = func(*args, **kwargs)

            return output

        return wrapper

    return decorator

def check_output(
    schema: DataFrameSchema, df_name: Optional[str] = None, raise_df_undefined: bool = True, pass_columns: bool = False
):
    """Decorator to validate output schema of decorated function.

    Args:
        schema: Schema to validate
        df_name (optional): name of output arg to validate
        raise_df_undefined: if set, ensures error is thrown if `df_name` is not defined
        pass_cols (optional): boolean indicating whether cols should be passed into callable
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            if not (type_ := typing.get_type_hints(func).get("return")):
                raise TypeError("No output typehint specified!")

            if df_name:
                if not typing.get_origin(typing.Dict) == dict:
                    raise TypeError("Specified df_name arg, but function output typehint is not dict.")

                type_ = typing.get_args(type_)[1]

            df_schema = schema.build_for_type(type_)

            if pass_columns:
                output = func(*args, **kwargs, cols=list(schema.columns.keys()))
            else:
                output = func(*args, **kwargs)

            if df_name is not None:
                df = output.get(df_name)

                if df is None and raise_df_undefined:
                    raise RuntimeError(f"df {df_name} not found!")
            else:
                df = output

            if df is not None:
                try:
                    df_schema.validate(df, lazy=False)
                except pa.errors.SchemaError as e:
                    _handle_schema_error("check_output", func, df_schema, df, e)

            return output

        return wrapper

    return decorator

```

A node file

```python
@check_output(
    schema=DataFrameSchema(
        columns={
            "source": Column(T.StringType(), nullable=False),
            "target": Column(T.StringType(), nullable=False),
            "source_embedding": Column(T.ArrayType(T.FloatType()), nullable=False),
            "target_embedding": Column(T.ArrayType(T.FloatType()), nullable=False),
        },
        unique=["source", "target"],
    )
)
def attach_embeddings(
    pairs_df: ps.DataFrame,
    nodes: ps.DataFrame,
) -> ps.DataFrame:

    # logic here ...
```

Reviewer 2: @lvijnck Is there any reason you can't just use the Pandera-provided `check_types` function decorator?

Cc @rashidakanchwala with the Pydantic integration leveraging type hints, maybe this is a way to go?

            "companies": companies_schema,
            "shuttles": shuttles_schema,
            "reviews": reviews_schema,
        }

    @hook_impl
    def before_node_run(
        self,
        node,
        catalog: DataCatalog,
        inputs: Dict[str, Any],
        is_async: bool,
    ) -> Dict[str, Any]:
        """Validate and coerce input data before node execution."""
        validated_inputs = {}

        for input_name, data in inputs.items():
            schema = self.raw_schemas.get(input_name)

>             if schema is None or not isinstance(data, pd.DataFrame):

Reviewer 2: Why is the `isinstance()` check required?

                logger.info(f"✓ Validation passed for '{input_name}'")
                validated_inputs[input_name] = validated_data
            except (SchemaError, SchemaErrors) as e:
                logger.error("Validation failed for '%s': %s", input_name, e)
                raise

        return validated_inputs
```

Finally, register the hook in `src/spaceflights_pandera/settings.py`:

```python
"""Project settings."""
from spaceflights_pandera.hooks.validation import PanderaValidationHooks

HOOKS = (PanderaValidationHooks(),)
```

From this point, when you execute `kedro run` you will see the validation logs:

```console
[10/01/25 11:10:13] INFO     Validating 'companies' for node 'preprocess_companies_node'
                    INFO     ✓ Validation passed for 'companies'
                    INFO     Validating 'shuttles' for node 'preprocess_shuttles_node'
                    INFO     ✓ Validation passed for 'shuttles'
                    INFO     Validating 'reviews' for node 'create_model_input_table_node'
                    INFO     ✓ Validation passed for 'reviews'
```

### What happens when validation fails?

> Let's make one schema rule deliberately too strict so you can see Pandera in action.

Reviewer 2: ```suggestion
Let's make one schema rule deliberately fail so you can see Pandera in action.
```
Nit: I think "too strict" would be more like if it was the wrong type but still usable; arbitrarily high value sounds more like deliberately failing. :)

In `src/spaceflights_pandera/schemas/raw.py`, change the `total_fleet_count` column check:

```diff
- checks=Check.greater_than_or_equal_to(0),
+ checks=Check.greater_than(100),  # intentionally too strict
```

Now run the pipeline again:

```bash
kedro run
```

You should see Pandera raise a `SchemaErrors` exception, reporting exactly which rows failed:

```console
SchemaErrors: Column 'total_fleet_count' failed check 'greater_than_or_equal_to(100)'
failure cases: [1.0, 2.0, 5.0, ...]
```

Because the hook uses `lazy=True`, Pandera collects all errors at once, making it easy to spot every problem.

> !!! note
>     Reset the schema back to the correct check (`greater_than_or_equal_to(0)`) before continuing with the rest of the tutorial.

Reviewer 2: ```suggestion
Reset the schema back to the correct check (`greater_than_or_equal_to(0)`) before continuing with the rest of the tutorial.
```

Nit: I don't know if this needs to be a callout.

> ### Controlling validation behavior with environment variables

Reviewer 2: This seems like a good, more advanced tip.

If we're covering things like this, I think it's also worth covering caching validation results (i.e. if a dataset is an input to 10 nodes, it's only validated once).

Reviewer 2: Also, looking below, is there a reason why this doesn't also fall under the advanced usage section?

                logger.info(f"✓ Validation passed for '{input_name}'")
                validated_inputs[input_name] = validated_data
             except (SchemaError, SchemaErrors) as e:
                logger.error("Validation failed for '%s': %s", input_name, e)
-                raise
+                if self.fail_on_error:
+                    raise
+                else:
+                    # Exploration mode: continue with original (unvalidated) data
+                    logger.warning(
+                        "Continuing despite validation errors because "
+                        "VALIDATION_FAIL_ON_ERROR=false"
+                    )
+                    validated_inputs[input_name] = data

-        return validated_inputs
+        return {**inputs, **validated_inputs}

```

Now you can control validation behavior:

```bash
# Disable validation entirely (for quick debugging)
export VALIDATION_ENABLED=false
kedro run

# Log errors but don't stop pipeline (for data exploration)
export VALIDATION_FAIL_ON_ERROR=false
kedro run

# Fail on first error (for faster CI/CD)
export VALIDATION_LAZY=false
kedro run
```

> ### Validating intermediate datasets

Reviewer 2: Does it add additional value to go through this?

Add this helper method to your `PanderaValidationHooks` class:

```python
def _handle_validation_error(self, dataset_name: str, error: SchemaError) -> None:
    """Parse and log validation errors in a readable format."""
    logger.error(f"{'='*60}")
    logger.error(f"VALIDATION FAILED: {dataset_name}")
    logger.error(f"{'='*60}")

    # Extract key error information
    if hasattr(error, 'failure_cases') and error.failure_cases is not None:
        logger.error(f"\nFailed checks summary:")
        logger.error(f"\n{error.failure_cases.to_string()}")

    logger.error(f"\nFull error:\n{error}")
    logger.error(f"{'='*60}")

    if self.fail_on_error:
        raise
```

Then use it in your validation logic:

```python
try:
    validated_data = schema.validate(data, lazy=schema.lazy)
    logger.info(f"✓ Validation passed for '{input_name}'")
    validated_inputs[input_name] = validated_data
except SchemaError as e:
    self._handle_validation_error(input_name, e)
    validated_inputs[input_name] = data
```

### DataFrame-level checks for business rules

> So far, our intermediate schema only checked column types and uniqueness.

Reviewer 2: But there was use of built-in `Check`s before that was more than just column types and uniqueness. Not sure what is "business rules;" maybe this is more custom checks?

That said, that's also covered in Pandera documentation, so not sure if need to cover all the types of checks unless the integration is different.

Author: This use case shows business-rule validation — logic involving multiple columns based on domain knowledge. You are right it doesn’t introduce new Pandera functionality. If others feel it overlaps with existing Pandera docs, I’m happy to remove it.

+    """Business rule: Type F5 shuttles cannot have more than 16 passengers."""
+    is_f5 = df["shuttle_type"] == "Type F5"
+    has_valid_capacity = df["passenger_capacity"] <= 16
+    return (~is_f5 | has_valid_capacity).all()
+

preprocessed_shuttles_schema = DataFrameSchema(
    columns={
        "id": Column(nullable=False, unique=True),
+        "shuttle_type": Column(str, nullable=True),
        "passenger_capacity": Column(nullable=False),
    },
+    checks=[
+        Check(
+            check_f5_shuttle_capacity,
+            error="Type F5 shuttles cannot have more than 16 passengers"
+        ),
+    ],
    strict=False,
    coerce=False,  # Data already cleaned, don't coerce
    name="preprocessed_shuttles",
)
```

!!! tip
    **Distribution checks and hypothesis testing**
    Beyond simple rules, Pandera supports **statistical hypothesis tests** (e.g. two-sample tests) to validate whether two datasets come from the same distribution.
    This is particularly useful for detecting **data drift** between training and serving environments.
    See [Hypothesis Testing](https://pandera.readthedocs.io/en/stable/hypothesis.html).

> ### Driving validation with Kedro parameters

Reviewer 2: Interesting use case, didn't think about it before

        })

        with pytest.raises(SchemaErrors) as exc_info:
            companies_schema.validate(invalid_data, lazy=True)

        assert "iata_approved" in str(exc_info.value)

    def test_companies_schema_rejects_negative_fleet(self):
        """Test companies schema rejects negative fleet counts."""
        invalid_data = pd.DataFrame({
            "id": [1],
            "company_rating": ["100%"],
            "company_location": ["USA"],
            "total_fleet_count": [-5.0],  # Must be >= 0
            "iata_approved": ["t"],
        })

        with pytest.raises(SchemaErrors):
            companies_schema.validate(invalid_data, lazy=True)
```

Run your tests:

```bash
python -m pytest src/spaceflights_pandera/schemas/test_schemas.py -v
```

## Further reading

* [Pandera documentation](https://pandera.readthedocs.io/)
* [Kedro hooks documentation](https://docs.kedro.org/en/stable/hooks/introduction.html)
* [Lazy validation](https://pandera.readthedocs.io/en/stable/lazy_validation.html)
* [Type coercion](https://pandera.readthedocs.io/en/stable/dtype_validation.html)
* [Custom checks](https://pandera.readthedocs.io/en/stable/checks.html)

> * [Multi-backend support](https://pandera.readthedocs.io/en/stable/ecosystem.html) - Pandera also validates data in Dask, Polars, Modin, and PySpark

Reviewer 2: Ibis also.

> For a declarative approach to validation using catalog metadata, see the community-maintained [kedro-pandera plugin](https://github.com/Galileo-Galilei/kedro-pandera).

Reviewer 2: It's not really maintained right now AFAIK. 😅
